### PR TITLE
Fix Qwen3-VL packed RoPE offsets and MoE export merging

### PR DESCRIPTION
**Pull request by yaoyu-33**
## Summary
- handle packed sequence attention masking when building Qwen3-VL mRoPE indices to fix packed seq logprob divergence (#1776)
- pass packed_seq_params through model to rope utils and use fp32 position ids for qwen3 mrope

Results can be verified via
```
uv run python -m torch.distributed.run --nproc_per_node=8 \
   examples/conversion/hf_to_megatron_generate_vlm.py --hf_model_path="Qwen/Qwen3-VL-30B-A3B-Instruct" \
   --image_path="https://qianwen-res.oss-cn-beijing.aliyuncs.com/Qwen-VL/assets/demo.jpeg" --prompt="Describe this image." \
   --ep 8

uv run python -m torch.distributed.run --nproc_per_node=8 examples/conversion/hf_megatron_roundtrip_multi_gpu.py \
      --hf-model-id Qwen/Qwen3-VL-30B-A3B-Instruct  --ep 2 --tp 2 --pp 2
```

### CODE REVIEW

File: src/megatron/bridge/utils/common_utils.py [hunk 1]
```diff
@@ -267,6 +268,59 @@ def extract_expert_number_from_param(param_name: str) -> int:
...
+        return converted_weights_dict
+
+    assert len(converted_weights_dict) == 1, (
+        f"There should be only one key in the converted_weights_dict, got keys: {converted_weights_dict.keys()}"
+    )
+
+    for key, value in converted_weights_dict.items():
+        cache = hf_weights_cache.setdefault(key, {})
+
+        if ep_size == 1:
```
Comment (lostkevin) [inline src/megatron/bridge/utils/common_utils.py]: always False?

File: src/megatron/bridge/models/qwen_vl/qwen3_vl_bridge.py [hunk 2]
```diff
@@ -359,38 +360,89 @@ def mapping_registry(self) -> MegatronMappingRegistry:
...
+            return {}
+
+        fused: Dict[str, torch.Tensor] = {}
+
+        for name, tensor in converted.items():
+            if name.endswith(".gate"):
+                base_name = name[: -len(".gate")]
+                gate_tensor = tensor
+                up_tensor = converted.get(f"{base_name}.up")
+                if up_tensor is None:
```
Comment (lostkevin) [inline src/megatron/bridge/models/qwen_vl/qwen3_vl_bridge.py]: I'm curious that when up_tensor will be None

File: src/megatron/bridge/models/qwen_vl/qwen3_vl_bridge.py [hunk 3]
```diff
@@ -359,38 +360,89 @@ def mapping_registry(self) -> MegatronMappingRegistry:
...
+            hf_weights_cache=self.hf_weights_cache,
+        )
+
 
 class ExpertMLPDownProjMapping(AutoMapping):
     """Mapping for expert MLP down projection weights between HF and Megatron formats."""
 
+    def __init__(self, *args, **kwargs):
+        # Down projection weights need a transpose; rely on AutoMapping permutation support.
+        permute_dims = kwargs.pop("permute_dims", (1, 0))
```
Comment (lostkevin) [inline src/megatron/bridge/models/qwen_vl/qwen3_vl_bridge.py]: Will bias mapping call torch.permute in AutoMapping?

Comment (yaoyu-33) [reply src/megatron/bridge/models/qwen_vl/qwen3_vl_bridge.py]: in this case, ExpertMLPDownProjMapping doesn't have bias in the mapping definition. But if it does, it will call permute as well.

File: src/megatron/bridge/models/qwen_vl/qwen3_vl_bridge.py (hunk 2 above)
Comment (yaoyu-33) [reply src/megatron/bridge/models/qwen_vl/qwen3_vl_bridge.py]: moved to a assert. it shouldn't happend.

File: src/megatron/bridge/utils/common_utils.py (hunk 1 above)
Comment (yaoyu-33) [reply src/megatron/bridge/utils/common_utils.py]: why always False?

Comment (lostkevin) [reply src/megatron/bridge/utils/common_utils.py]: ep_size 1 will return at L286 in advance

File: src/megatron/bridge/models/qwen_vl/qwen3_vl_bridge.py (hunk 3 above)
Comment (lostkevin) [reply src/megatron/bridge/models/qwen_vl/qwen3_vl_bridge.py]: I see，but bias is an 1-d tensor. Will permute raise error if this happens？（GPT-OSS has biases，therefore I don’t know if it works on that model）

File: src/megatron/bridge/utils/common_utils.py (hunk 1 above)
Comment (yaoyu-33) [reply src/megatron/bridge/utils/common_utils.py]: lol my bad.

File: src/megatron/bridge/models/qwen_vl/qwen3_vl_bridge.py (hunk 3 above)
Comment (yaoyu-33) [reply src/megatron/bridge/models/qwen_vl/qwen3_vl_bridge.py]: yeah it will raise an error. We haven't synchronized this projection part here. We can sync later and yeah maybe this need to changed. (or we update permute logic)

**Review by lostkevin — APPROVED** [review]

File: src/megatron/bridge/models/qwen_vl/qwen3_vl_bridge.py [hunk 4]
```diff
@@ -429,20 +476,18 @@ def megatron_to_hf(self, megatron_weights: torch.Tensor, megatron_module: nn.Mod
...
+                # [ep_size, mlp_in, up_out]
+                up_tensor = converted.get(f"{base_name}.up").transpose(1, 2).contiguous()
                 assert up_tensor is not None
-                # Back to HF layout.
-                gate_tensor = gate_tensor.transpose(0, 1).contiguous()
-                up_tensor = up_tensor.transpose(0, 1).contiguous()
-                fused[base_name] = torch.cat([gate_tensor, up_tensor], dim=0)
+                # Back to HF fused layout: stack [gate; up] along dim 0.
+                # [ep_size, 2, mlp_in, gate_out/up_out]
+                fused[base_name] = torch.stack([gate_tensor, up_tensor], dim=1)
```
Comment (lostkevin) [inline src/megatron/bridge/models/qwen_vl/qwen3_vl_bridge.py]: why not concat here?

Comment (yaoyu-33) [reply src/megatron/bridge/models/qwen_vl/qwen3_vl_bridge.py]: if concat here needs to split later.
if concat we will have [gate_expert_1, up_expert_1, gate_expert_2, up_expert_2, ...], we kinda need to uninterleave it later

File: src/megatron/bridge/models/gpt_oss/gpt_oss_bridge.py [hunk 5]
```diff
@@ -240,10 +240,6 @@ def megatron_to_hf(self, megatron_weights: torch.Tensor, megatron_module: nn.Mod
             megatron_weights = megatron_weights.transpose(0, 1)
         return super().megatron_to_hf(megatron_weights.contiguous(), megatron_module)
 
-    def _validate_patterns(self, *args, **kwargs):
-        # allow number of wildcards to mismatch in this mapping
-        pass
-
```
Comment (cuichenx) [inline src/megatron/bridge/models/gpt_oss/gpt_oss_bridge.py:246]: This way gpt-oss would print a warning statement but it should be ignored by the user, right? It might be counter intuitive for new users
